hedgehog/src/Hedgehog/Internal/Property.hs

```diff
@@ -331,6 +343,159 @@ newtype ShrinkCount =
   ShrinkCount Int
   deriving (Eq, Ord, Show, Num, Enum, Real, Integral)
 
+-- | Where to start running a property's tests.
+--
+data Skip =
+  -- | Don't skip anything.
+  --
+    SkipNothing
+
+  -- | Skip to a specific test number. If it fails, shrink as normal. If it
+  --   passes, move on to the next test. Coverage checks are disabled.
+  --
+  | SkipToTest TestCount
+
+  -- | Skip to a specific test number and shrink state. If it fails, stop
+  --   without shrinking further. If it passes, the property will pass without
+  --   running any more tests.
+  --
+  --   Due to implementation details, all intermediate shrink states - those on
+  --   the direct path from the original test input to the target state - will
+  --   be tested too, and their results discarded.
+  --
+  | SkipToShrink TestCount ShrinkPath
+  deriving (Eq, Ord, Show, Lift)
+
+-- | We use this instance to support usage like
+--
+-- @
+--   withTests "3:aB"
+-- @
+--
+--   It throws an error if the input is not a valid compressed 'Skip'.
+--
+instance IsString Skip where
+  fromString s =
+    case skipDecompress s of
+      Nothing ->
+        error $ "Not a valid Skip: " ++ s
+      Just skip ->
+        skip
+
+-- | The path taken to reach a shrink state.
+--
+--   This is in reverse order, most recent path element first, for efficient
+--   appending.
+--
+newtype ShrinkPath =
+  ShrinkPath [Int]
+  deriving (Eq, Ord, Show, Lift)
```

Comment:
Being in reverse order is awkward, and having to know this fact while interfacing with it is awkward too. I think there are two independent questions here:

First, what should `ShrinkPath` wrap? Some options:

* `Dlist Int`. Requires [dlist](https://www.stackage.org/haddock/lts-19.9/dlist-1.0/Data-DList.html) as a dependency, but offers O(1) snoc.
* `[Int] -> [Int]`, which is what a `DList Int` is under the hood.
* `[Int]` as currently, but keep it in order. snoc becomes O(n), which means constructing the full list is O(n^2). Maybe n is small enough that that doesn't matter? I've seen it be in the hundreds, so my guess is this is worth some complexity to avoid, but I don't have strong intuitions about that.
* Keep as-is.

Second, how should it be exposed and interfaced with? Some options:

* Expose constructor as currently.
* Hide the constructor, offer functions like `shrinkPathToList`, `shrinkPathFromList`, `shrinkPathSnoc`.
* Hide the constructor, put it in its own module. Name the functions simply `toList`, `fromList`, `snoc`.

Reply:
I'd give it a try with the `[Int]` in the right order as this is the config API.

It looks like the the snoc only occurs in `takeSmallest` so you could always switch to a reversed list only inside that function if you think it could be an issue.

In what situations do think people will interact with `ShrinkPath`? That will guide what we should expose. You can see `TestLimit` and the like do not have their constructors exposed.

I'm always cautious to expose anything that's not needed as I'm pretty strict with maintaining backwards compatibility and making upgrades easy on users. They always have the option to go to the internal modules for advanced use cases.

Reply:
Ah yeah, I've put it in order and gone from three uses of `reverse` to just two, in `takeSmallest`. I agree that's better than before.

I wasn't actually thinking of exposing the constructor to users, I can't think of a reason they'd use it. I was more thinking of how other internal parts would interface with it.